tests/ctst/steps/bucket-policies/common.ts

```diff
@@ -0,0 +1,475 @@
+                effect = AuthorizationType.DENY;
+            }
+            resources = [
+                `arn:aws:s3:::${bucketName}`,
+                `arn:aws:s3:::${bucketName}/*`,
+            ];
+        } else {
+            authzConfiguration.Resource = AuthorizationType.IMPLICIT_DENY;
+            // Effect is ALLOW on purpose, to ensure we properly handle implicit denies
+            effect = AuthorizationType.ALLOW;
+            resources = [
+                `arn:aws:s3:::${bucketName}`,
+                `arn:aws:s3:::${bucketName}/*`,
+            ];
+        }
+        if (action.excludePermissionOnBucketObjects) {
+            resources.pop();
+        }
+        if (action.excludePermissionsOnBucket) {
+            resources.shift();
+        }
+    } else {
+        authzConfiguration.Resource = AuthorizationType.NO_RESOURCE;
+        return;
+    }
+    const expectFromConditions = this.getSaved<AuthorizationType>('expectFromConditions');
+    if (expectFromConditions) {
+        authzConfiguration.Resource = expectFromConditions;
+    }
+    this.addToSaved('authzConfiguration', authzConfiguration);
+    const currentIdentityArn = this.getSaved<string>('identityArn');
+    let principal = currentIdentityArn;
+    const identityType = this.getSaved<string>('identityType') as EntityType;
+    if (identityType === EntityType.ASSUME_ROLE_USER
+        || identityType === EntityType.ASSUME_ROLE_USER_CROSS_ACCOUNT
+        || identityType === EntityType.DATA_CONSUMER) {
+        principal = '*';
+    }
+    if (!applies) {
+        principal = `${currentIdentityArn}badname`;
+    }
+    const basePolicy = {
+        Version: '2012-10-17',
+        Statement: [
+            {
+                Effect: effect,
+                Action: action.permissions,
+                Resource: resources,
+                Principal: {
+                    AWS: principal,
+                },
+                Condition: {},
+            },
+        ],
+    };
+    const conditionForPolicy = this.getSaved('conditionForPolicy');
+    if (conditionForPolicy) {
+        basePolicy.Statement[0].Condition = conditionForPolicy;
+    }
+    const result = await S3.putBucketPolicy({
+        bucket: this.getSaved<string>('bucketName'),
+        policy: JSON.stringify(basePolicy),
+    });
+    assert.ifError(result.stderr || result.err);
+});
+
+Given('an environment setup for the API', async function (this: Zenko) {
+    const action = this.getSaved<ActionPermissionsType>('currentAction');
+    if (!action.needsSetup) {
+        return;
+    }
+    // Create an IAM policy with full S3 permission on any bucket
+    // and attach it to the current identity
+    this.setAuthMode('base_account');
+    const basePolicy = {
+        Version: '2012-10-17',
+        Statement: [
+            {
+                Effect: 'Allow',
+                Action: '*',
+                Resource: '*',
+            },
+        ],
+    };
```

Comment:
Just to be sure, this is only used in the context of making sure we get Denies when we expect them from one of the policies, correct?

Reply:
The environment setup step is an optional step where we setup additional objects/configuration in the case some APIs are returning e.g. `AccesDeny` or fail before reaching the authz checks. In such cases we cannot test if we authorize the right way, so we need to setup things.
This specific code creates a temporary policy as the base account to grant the current identity (IAM User or Role) with all permissions to setup the environment. The reason is also because some of our APIs are requiring that the initiator is the same identity as the one completing a MPU. That's why the actual setup is:

- Account grants the identity with all permissions
- The identity setups the environment
- The base account removes the permissions from the identity

At the end we continue with the test scenario as usual